src/realm/sync/client.cpp

```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
 {
     REALM_ASSERT(!m_finalized);
     REALM_ASSERT(m_sess);
 
     if (!m_progress_handler)
         return;
 
-    std::uint_fast64_t downloaded_bytes = 0;
-    std::uint_fast64_t downloadable_bytes = 0;
-    std::uint_fast64_t uploaded_bytes = 0;
-    std::uint_fast64_t uploadable_bytes = 0;
-    std::uint_fast64_t snapshot_version = 0;
-    ClientHistory::get_upload_download_bytes(m_db.get(), downloaded_bytes, downloadable_bytes, uploaded_bytes,
-                                             uploadable_bytes, snapshot_version);
+    // Ignore progress messages from before we first receive a DOWNLOAD message
+    if (!m_reliable_download_progress)
+        return;
+
+    ReportedProgress p = m_reported_progress;
+    ClientHistory::get_upload_download_bytes(m_db.get(), p.downloaded, p.downloadable, p.uploaded, p.uploadable,
+                                             p.snapshot);
 
     // If this progress notification was triggered by a commit being made we
     // only want to send it if the uploadable bytes has actually increased,
     // and not if it was an empty commit.
-    if (only_if_new_uploadable_data && m_last_reported_uploadable_bytes == uploadable_bytes)
+    if (only_if_new_uploadable_data && m_reported_progress.uploadable == p.uploadable)
         return;
-    m_last_reported_uploadable_bytes = uploadable_bytes;
 
     // uploadable_bytes is uploaded + remaining to upload, while downloadable_bytes
     // is only the remaining to download. This is confusing, so make them use
     // the same units.
-    std::uint_fast64_t total_bytes = downloaded_bytes + downloadable_bytes;
-
-    m_sess->logger.debug("Progress handler called, downloaded = %1, "
-                         "downloadable(total) = %2, uploaded = %3, "
-                         "uploadable = %4, reliable_download_progress = %5, "
-                         "snapshot version = %6",
-                         downloaded_bytes, total_bytes, uploaded_bytes, uploadable_bytes,
-                         m_reliable_download_progress, snapshot_version);
-
-    // FIXME: Why is this boolean status communicated to the application as
-    // a 64-bit integer? Also, the name `progress_version` is confusing.
-    std::uint_fast64_t progress_version = (m_reliable_download_progress ? 1 : 0);
-    m_progress_handler(downloaded_bytes, total_bytes, uploaded_bytes, uploadable_bytes, progress_version,
-                       snapshot_version);
+    p.downloadable += p.downloaded;
+
+    double upload_estimate = 1.0, download_estimate = 1.0;
+
+    if ((!is_completed || is_download) && p.uploaded != p.uploadable) {
+        if (p.uploadable > p.final_uploaded)
+            upload_estimate = (p.uploaded - p.final_uploaded) / double(p.uploadable - p.final_uploaded);
+    }
+
+    // download estimate only known for flx
+    if (m_download_estimate) {
+        download_estimate = *m_download_estimate;
+
+        // ... and transient bytes are only relevant with bootstrap store
```

Comment:
we should then name it accordingly (i.e, bootstrap_downloaded_bytes), since this is a bit confusing